design/pipeline-caching.md

```diff
@@ -0,0 +1,366 @@
+# Pipeline Caching
+
+*** Status: Work in Progress ***
+
+We have observed the need for Azure Pipelines to implement a caching mechanism that allows the outputs of steps within the pipeline to be skipped if a suitable cached version of the output already exists.
+
+Provided the cost of determining a cache hit and acquiring the contents of the cache is cheaper than the cost of producing the output again from scratch then the pipeline's performance will be increased.
+
+Caching can introduce many complexities into build pipelines and our goal is to strike the balance between "correct" and "good enough" where "good enough" allows us to provide performance benefits without a disproporionate amount of effort on behalf of the user.
+
+## Proposal
+
+This proposal contains a number of interelated elements. I've tried to build it up sequentially but sometimes you'll need to read further along to understand why something is the way it is.
+
+### The basics & task integration
+
+When work is done on an agent as part of a pipeline, it is represented as a _job_ when a set of _steps_ which are implemented by tasks. For many pipelines the YAML syntax obscures the underlying "tasky" nature of Azure Pipelines, but they are there nonetheless. As a result I'm starting with how pipeline caching will integrate with the underlying task infrastructure.
+
+Detecting a cache hit or miss will be performend by the agent worker. As the worker prepares to execute each task it will check to see whether any given task has caching enabled and if so it will determine the key and check for a cache hit. If a hit is detected it will set about restoring the cached content into the pipeline execution environment. Exactly how that content is restored will be discussed a bit later.
+
+The benefit of this approach is that caching can be applied to all existing tasks without any additional work by the task author. Support for caching can be enhanced to provide sensible defaults (e.g. what files to consider for cache key input).
```

Comment:
I don't think this is going to be correct in many cases.  For the most part you will likely still want to run the task to normalize the cache to make sure you have the right versions as there are a lot of cases where you might choose a key like branchname and that would give you mostly the right files and then you only have to download the delta from the package source.